### PR TITLE
✨ Clusterctl move with clusterclass

### LINKED ISSUE #5445: Enhance `clusterctl move` to work with clusterclass and managed topologies
**Description**

Add support to the `clusterctl move` command to also move the ClusterClass (and associated objects) to the target cluster. 

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind feature
/area topology
/area clusterctl

/assign @ykakarap

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:
This PRs add ClusterClass support to `clusterctl move` command.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5445

### CODE REVIEW

File: internal/builder/builders.go [hunk 1]
```diff
@@ -313,8 +313,12 @@ func (m *MachineDeploymentClassBuilder) WithAnnotations(annotations map[string]s
 
 // Build creates a full MachineDeploymentClass object with the variables passed to the MachineDeploymentClassBuilder.
 func (m *MachineDeploymentClassBuilder) Build() *clusterv1.MachineDeploymentClass {
+	class := m.class
+	if class == "" {
+		class = m.name
+	}
```
[the remark this answers is not in the document]
Comment (ykakarap) [reply internal/builder/builders.go]: done.

**Comment by vincepri** [comment]
Can we unhold?

File: cmd/clusterctl/internal/test/fake_objects.go [hunk 2]
```diff
@@ -1371,3 +1377,136 @@ func FakeCRDList() []*apiextensionsv1.CustomResourceDefinition {
...
+func (f *FakeClusterClass) WithWorkerMachineDeploymentClasses(classes []*FakeMachineDeploymentClass) *FakeClusterClass {
+	f.workerMachineDeploymentClasses = classes
+	return f
+}
+
+func (f *FakeClusterClass) Objs() []client.Object {
+	// objMap map where the key is the object to which the owner reference to the cluster class should be added
+	// and the value dictates if the onwner ref needs to be added.
+	// This map also dual functions as a way to de-duplicate and template objects that are reused.
+	objMap := map[client.Object]bool{}
```
Comment (sbueringer) [inline cmd/clusterctl/internal/test/fake_objects.go:1421]: Can a client.Object be used as a map key?

I wonder if it's comparable.

> As mentioned earlier, map keys may be of any type that is comparable. The language spec defines this precisely, but in short, comparable types are boolean, numeric, string, pointer, channel, and interface types, and structs or arrays that contain only those types. Notably absent from the list are slices, maps, and functions; these types cannot be compared using ==, and may not be used as map keys.
> https://go.dev/blog/maps#TOC_5.

Comment (sbueringer) [reply cmd/clusterctl/internal/test/fake_objects.go:1421]: I assume it is safe as the only types which could be duplicate are Unstructured and Unstructured is essentially only nested maps?

I wonder what would happen with regular types (but we don't have to care about them here)

**Review by sbueringer — COMMENTED** [review]
Looks good to me. 

A few questions to confirm my understanding of `clusterctl move`.

`clusterctl move` currently only moves entire namespaces, correct? So we don't have to care about which Cluster is using which ClusterClass and if the ClusterClass can already be moved, etc. ...

I wonder if we should have an e2e test for clusterctl move at some point (not in this PR). Out of curiosity did you test it manually to make sure it works and we're not missing anything?

File: cmd/clusterctl/internal/test/fake_objects.go (hunk 2 above)
Comment (ykakarap) [reply cmd/clusterctl/internal/test/fake_objects.go:1421]: Yup, since this is only for the test builder with just unstructured objects it works.

**Comment by ykakarap** [comment]
> clusterctl move currently only moves entire namespaces, correct? So we don't have to care about which Cluster is using which ClusterClass and if the ClusterClass can already be moved, etc. ...

Yes, clsuterctl move operated on one namespace. Since ClusterClasses are also namespaced and a cluster can only use a clusterclass in the same namespace we can be sure that if a cluster is moving then the associated ClusterClass is also moving. To add to that ClusterClass is treated as a root of the graph. So ClusterClass along with all the associated templates (nodes in the clusterclass graph) will move along with it. 

>I assume this PR enable clusterctl move, backup and restore with ClusterClass?

Yes. Backup and restore also uses the same underlying tooling as move. So once this PR will also support backup and restore.

>I wonder if we should have an e2e test for clusterctl move at some point (not in this PR). Out of curiosity, did you test it manually to make sure it works and we're not missing anything?

Yes, I am already working on adding e2e tests for clusterctl move. I tested this a few times manually. After solving [this](https://kubernetes.slack.com/archives/C8TSNPY4T/p1635956040309100) I did not hit any other issues.

**Comment by sbueringer** [comment]
Great, thx!
/lgtm

**Pull request merged**